### PR TITLE
prevent !grep denial-of-service

### PR DESCRIPTION
**Pull request by fwilson42**

### CODE REVIEW

**Comment by pefoley2** [comment]
Needs an extra .count() query to get the number of instances.

**Comment by fwilson42** [comment]
@pefoley2 that would do a full table scan which is the problem this tries to prevent -- speed/feature tradeoff

**Comment by pefoley2** [comment]
The issue is loading the result of .all() into ram.

**Comment by fwilson42** [comment]
in that case

**Comment by fwilson42** [comment]
how about this

File: cslbot/commands/grep.py [hunk 1]
```diff
@@ -50,11 +50,13 @@ def cmd(send, msg, args):
         query = query.filter(Log.msg.ilike('%%%s%%' % escape(term)))
     else:
         query = query.filter(Log.msg.like('%%%s%%' % escape(term)))
-    rows = query.order_by(Log.time.desc()).all()
+    query = query.order_by(Log.time.desc())
+    rows = query.limit(1).all()
```
Comment (pefoley2) [inline cslbot/commands/grep.py]: .first()

File: cslbot/commands/grep.py [hunk 2]
```diff
@@ -50,11 +50,13 @@ def cmd(send, msg, args):
...
+    count = query.count()
     if rows:
         row = rows[0]
```
Comment (pefoley2) [inline cslbot/commands/grep.py]: Don't need a list anymore.

**Comment by pefoley2** [comment]
Have you tested that this change works?

**Comment by fwilson42** [comment]
absolutely not

**Comment by pefoley2** [comment]
Please add a test for this feature then.

**Comment by fwilson42** [comment]
no

**Pull request merged**